File: CyScat/Scattering_Code/transper.jl
"""
transper - Translation matrix for periodic structures
Coder : Curtis Jin
Contact : jsirius@umich.edu
Description : Translation matrix calculation for periodic case
              using spectral or spatial methods with Shanks transformation

Julia translation
"""

using SpecialFunctions
using LinearAlgebra

include("ky.jl")
include("modified_epsilon_shanks.jl")

"""
    transper(clocob, cmob, clocso, cmso, period, lambda_val, phiinc,
             epsloc, kshanks, epsseries, nrepeat, spectral, spec_cond)

Calculate translation coefficient for periodic structures.

# Arguments
- `clocob`: Observer cylinder location (1x2)
- `cmob`: Observer mode number
- `clocso`: Source cylinder location (1x2)
- `cmso`: Source mode number
- `period`: Periodic spacing (negative for non-periodic)
- `lambda_val`: Wavelength
- `phiinc`: Incident angle
- `epsloc`: Spatial tolerance for close cylinders
- `kshanks`: Shanks transformation order
- `epsseries`: Convergence tolerance
- `nrepeat`: Number of times convergence must be met
- `spectral`: Use spectral method if positive
- `spec_cond`: Y distance threshold for switching to spatial method

# Returns
- `t`: Translation coefficient
"""
function transper(clocob, cmob, clocso, cmso, period, lambda_val, phiinc,
                  epsloc, kshanks, epsseries, nrepeat, spectral, spec_cond)
    k = 2π / lambda_val - 1im * 1e-14
    jmax = 3000
    spectralinternal = spectral

    rv = clocob .- clocso
    r = norm(rv)
    x, y = rv[1], rv[2]

    if period < 0
        # Non-periodic case
        if r < epsloc
            t = 0.0 + 0.0im
        else
            t = trans(clocob, cmob, clocso, cmso, lambda_val)
        end
        return t
    end

    # Periodic case
    # Spectral condition checker
    if abs(y) < spec_cond
        spectralinternal = -1
        println("Calculating Spatial Sum!")
    end

    # Shanks Transformation Initialization
    ashankseps1 = zeros(ComplexF64, kshanks + 2)
    ashankseps2 = zeros(ComplexF64, kshanks + 2)
    if kshanks > 0
        ashankseps1[1] = Inf
        ashankseps2[1] = Inf
    end

    ts = 1.0 + 0.0im

    if spectralinternal < 0
        # Spatial case
        if r < epsloc
            t = 0.0 + 0.0im
        else
            t = trans(clocob, cmob, clocso, cmso, lambda_val)
        end
        if kshanks > 0
            ashankseps1[2] = t
        end
    else
        # Spectral case
        cm = -(cmob - cmso)
        kx0 = k * cos(phiinc)
        kx = kx0

        ky_val = ky(k, kx)
        if isa(ky_val, AbstractArray)
            ky_val = ky_val[1]
        end
        sign_y = y != 0 ? sign(y) : 1

        # Complex arcsin for evanescent modes
        angle = asin(Complex(kx / k))

        t = (Float64(sign_y)^cm *
             exp(-1im * kx * x - 1im * ky_val * abs(y) -
                 1im * cm * (sign_y * angle - π)) /
             ky_val * 2 / period)

        if kshanks > 0
            ashankseps1[2] = t
        end
    end

    # Looping
    j = 1
    irepeat = 0
    while irepeat < nrepeat && j < jmax
        # Sequence calculation
        if spectralinternal < 0
            # Non-spectral (spatial) method
            add = (exp(-1im * k * j * period * cos(phiinc)) *
                   trans(clocob, cmob, clocso .+ j .* [period, 0], cmso, lambda_val) +
                   exp(-1im * k * (-j) * period * cos(phiinc)) *
                   trans(clocob, cmob, clocso .- j .* [period, 0], cmso, lambda_val))
        else
            # Spectral method
            cm = -(cmob - cmso)
            kx0 = k * cos(phiinc)
            kxp = j * 2π / period + kx0
            kxm = -j * 2π / period + kx0

            ky_p = ky(k, kxp)
            ky_m = ky(k, kxm)
            if isa(ky_p, AbstractArray)
                ky_p = ky_p[1]
            end
            if isa(ky_m, AbstractArray)
                ky_m = ky_m[1]
            end

            sign_y = y != 0 ? sign(y) : 1

            # Complex arcsin for evanescent modes
            angle_p = asin(Complex(kxp / k))
            angle_m = asin(Complex(kxm / k))

            add = (Float64(sign_y)^cm * 2 / period *
                   (exp(-1im * kxp * x - 1im * ky_p * abs(y) -
                        1im * cm * (sign_y * angle_p - π)) / ky_p +
                    exp(-1im * kxm * x - 1im * ky_m * abs(y) -
                        1im * cm * (sign_y * angle_m - π)) / ky_m))
        end

        # Shanks Transformation & Convergence Checker
        told = t
        t = t + add

        if kshanks <= 0
            rerror = abs(t - told)
        else
            S = modified_epsilon_shanks(t, ashankseps1, ashankseps2)
            ashankseps2 = copy(ashankseps1)
            ashankseps1 = S

            if j <= 2 * kshanks + 1
                rerror = 1.0
            else
                tsold = ts
                ts = S[end]
                rerror = abs(ts - tsold)
            end
        end

        if rerror < epsseries
            irepeat = irepeat + 1
        else
            irepeat = 0
        end

        j = j + 1
    end

    if j == jmax
        println("j has reached jmax!! j = $j")
    end

    if kshanks > 0 && period > 0
        t = ts
    end

    return t
end


"""
    trans(clocob, cmob, clocso, cmso, lambda_val)

Calculate single translation coefficient (non-periodic).
"""
function trans(clocob, cmob, clocso, cmso, lambda_val)
    k = 2π / lambda_val
    rv = clocob .- clocso
    x, y = rv[1], rv[2]
    r = norm(rv)
    phip = atan(y, x)

    t = hankelh2(cmob - cmso, k * r) * exp(-1im * (cmob - cmso) * (phip - π))

    return t
end

File: python/CyScat/Scattering_Code/transper.py
"""
transper - Translation matrix for periodic structures
Coder : Curtis Jin
Contact : jsirius@umich.edu
Description : Translation matrix calculation for periodic case
              using spectral or spatial methods with Shanks transformation

Python translation
"""

import numpy as np
from scipy.special import hankel2
from .ky import ky
from .modified_epsilon_shanks import modified_epsilon_shanks


def transper(clocob, cmob, clocso, cmso, period, lambda_val, phiinc,
             epsloc, kshanks, epsseries, nrepeat, spectral, spec_cond):
    """
    Calculate translation coefficient for periodic structures.

    Parameters:
    -----------
    clocob : ndarray
        Observer cylinder location (1x2)
    cmob : int
        Observer mode number
    clocso : ndarray
        Source cylinder location (1x2)
    cmso : int
        Source mode number
    period : float
        Periodic spacing (negative for non-periodic)
    lambda_val : float
        Wavelength
    phiinc : float
        Incident angle
    epsloc : float
        Spatial tolerance for close cylinders
    kshanks : int
        Shanks transformation order
    epsseries : float
        Convergence tolerance
    nrepeat : int
        Number of times convergence must be met
    spectral : int
        Use spectral method if positive
    spec_cond : float
        Y distance threshold for switching to spatial method

    Returns:
    --------
    t : complex
        Translation coefficient
    """
    k = 2 * np.pi / lambda_val - 1j * 1e-14
    jmax = 3000
    spectralinternal = spectral

    rv = clocob - clocso
    r = np.linalg.norm(rv)
    x, y = rv[0], rv[1]

    if period < 0:
        # Non-periodic case
        if r < epsloc:
            t = 0
        else:
            t = trans(clocob, cmob, clocso, cmso, lambda_val)
        return t

    # Periodic case
    # Spectral condition checker
    if np.abs(y) < spec_cond:
        spectralinternal = -1
        print('Calculating Spatial Sum!')

    # Shanks Transformation Initialization
    if kshanks > 0:
        ashankseps1 = np.zeros(kshanks + 2, dtype=complex)
        ashankseps2 = np.zeros(kshanks + 2, dtype=complex)
        ashankseps1[0] = np.inf
        ashankseps2[0] = np.inf

    if spectralinternal < 0:
        # Spatial case
        if r < epsloc:
            t = 0
        else:
            t = trans(clocob, cmob, clocso, cmso, lambda_val)
        ashankseps1[1] = t
        ts = 1
    else:
        # Spectral case
        cm = -(cmob - cmso)
        kx0 = k * np.cos(phiinc)
        kx = kx0

        ky_val = ky(k, kx)
        sign_y = np.sign(y) if y != 0 else 1

        # Complex arcsin for evanescent modes
        angle = np.arcsin(kx / k + 0j)

        t = (sign_y ** cm *
             np.exp(-1j * kx * x - 1j * ky_val * np.abs(y) -
                    1j * cm * (sign_y * angle - np.pi)) /
             ky_val * 2 / period)

        ashankseps1[1] = t
        ts = 1

    # Looping
    j = 1
    irepeat = 0
    while irepeat < nrepeat and j < jmax:
        # Sequence calculation
        if spectralinternal < 0:
            # Non-spectral (spatial) method
            add = (np.exp(-1j * k * j * period * np.cos(phiinc)) *
                   trans(clocob, cmob, clocso + j * np.array([period, 0]), cmso, lambda_val) +
                   np.exp(-1j * k * (-j) * period * np.cos(phiinc)) *
                   trans(clocob, cmob, clocso - j * np.array([period, 0]), cmso, lambda_val))
        else:
            # Spectral method
            kxp = j * 2 * np.pi / period + kx0
            kxm = -j * 2 * np.pi / period + kx0

            ky_p = ky(k, kxp)
            ky_m = ky(k, kxm)

            # Complex arcsin for evanescent modes
            angle_p = np.arcsin(kxp / k + 0j)
            angle_m = np.arcsin(kxm / k + 0j)

            add = (sign_y ** cm * 2 / period *
                   (np.exp(-1j * kxp * x - 1j * ky_p * np.abs(y) -
                           1j * cm * (sign_y * angle_p - np.pi)) / ky_p +
                    np.exp(-1j * kxm * x - 1j * ky_m * np.abs(y) -
                           1j * cm * (sign_y * angle_m - np.pi)) / ky_m))

        # Shanks Transformation & Convergence Checker
        told = t
        t = t + add

        if kshanks <= 0:
            rerror = np.abs(t - told)
        else:
            S = modified_epsilon_shanks(t, ashankseps1, ashankseps2)
            ashankseps2 = ashankseps1.copy()
            ashankseps1 = S

            if j <= 2 * kshanks + 1:
                rerror = 1
            else:
                tsold = ts
                ts = S[-1]
                rerror = np.abs(ts - tsold)

        if rerror < epsseries:
            irepeat = irepeat + 1
        else:
            irepeat = 0

        # Increment
        j = j + 1

    if j == jmax:
        print(f'j has reached jmax!! j = {j}')

    if kshanks > 0 and period > 0:
        t = ts

    return t


def trans(clocob, cmob, clocso, cmso, lambda_val):
    """
    Calculate single translation coefficient (non-periodic).
    """
    k = 2 * np.pi / lambda_val
    rv = clocob - clocso
    x, y = rv[0], rv[1]
    r = np.linalg.norm(rv)
    phip = np.arctan2(y, x)

    t = hankel2(cmob - cmso, k * r) * np.exp(-1j * (cmob - cmso) * (phip - np.pi))

    return t
